espnex/models/transformer/embedding.py

```diff
@@ -0,0 +1,84 @@
+import math
+
+import numpy as np
+import flax.linen as nn
+from flax.linen import LayerNorm, Dense, Dropout, Module
+import jax
+import jax.numpy as jnp
+from flax.linen.module import merge_param
+from jax.random import bernoulli
+from jax import Array
+
+
+def sinusoidal_init(max_len=2048,
+                    min_scale=1.0,
+                    max_scale=10000.0,
+                    init_type='flax'):
+    """1D Sinusoidal Position Embedding Initializer. Copy from https://github.com/google/flax/tree/main/examples
+    Args:
+        max_len: maximum possible length for the input.
+        min_scale: float: minimum frequency-scale in sine grating.
+        max_scale: float: maximum frequency-scale in sine grating.
+    Returns:
+        output: init function returning `(1, max_len, d_feature)`
+    """
+
+    def init(key, shape, dtype=np.float32):
+        """Sinusoidal init."""
+        del key, dtype
+        d_feature = shape[-1]
+        pe = np.zeros((max_len, d_feature), dtype=np.float32)
+        position = np.arange(0, max_len)[:, np.newaxis]
+        scale_factor = -np.log(max_scale / min_scale) / (d_feature // 2 - 1)
+        div_term = min_scale * np.exp(np.arange(0, d_feature // 2) * scale_factor)
+        pe[:, :d_feature // 2] = np.sin(position * div_term)
+        pe[:, d_feature // 2: 2 * (d_feature // 2)] = np.cos(position * div_term)
+        pe = pe[np.newaxis, :, :]  # [1, max_len, d_feature]
+        return jnp.array(pe)
+
+    def init_espnet(key, shape, dtype=np.float32):
+        """Sinusoidal init."""
+        del key, dtype
+        d_feature = shape[-1]
+        pe = np.zeros((max_len, d_feature), dtype=np.float32)
+        position = np.arange(0, max_len)[:, np.newaxis]
+        scale_factor = -np.log(max_scale / min_scale) / d_feature
+        div_term = min_scale * np.exp(np.arange(0, d_feature, 2) * scale_factor)
+        pe[:, ::2] = np.sin(position * div_term)
+        if max_len % 2:
```

Comment:
This part is mainly copied from torch version. I add this line to check even/odd of `max_len` since it seems that ther was a bug when max_len is odd. (`pe[:, 1::2].shape != position.shape`) @simpleoier